### ISSUE TITLE
creating a second service instance causes stacktrace

### ISSUE THREAD

**Issue by Blackbaud-JasonCrowe** [epoelke/rds-service-broker#1]
using 1.5.7 ops manager and 1.5.6 elastic runtime

command:
CF_TRACE=true cf create-service hsdp-rds postgres-medium-dev pg-med-2

relevant output:
```
REQUEST: [2015-10-20T16:23:57Z]
POST /v2/service_instances?accepts_incomplete=true HTTP/1.1^M
Host: api.system.pivotal.blackbaudcloud.com^M
Accept: application/json^M
Authorization: [PRIVATE DATA HIDDEN]
Content-Type: application/json^M
User-Agent: go-cli 6.11.3-cebadc9 / linux^M
^M
{"name":"pg-med-2","space_guid":"ff2844e8-f063-4e53-bd01-636bfb839bc7","service_plan_guid":"9db73a23-0858-4768-a18e-5d8ab10d75af"}

RESPONSE: [2015-10-20T16:23:57Z]
HTTP/1.1 502 Bad Gateway^M
Content-Length: 1403^M
Connection: keep-alive^M
Content-Type: application/json;charset=utf-8^M
Date: Tue, 20 Oct 2015 16:23:57 GMT^M
Server: nginx^M
X-Cf-Requestid: 17a3172e-a382-4786-7832-99428bab932c^M
X-Content-Type-Options: nosniff^M
X-Vcap-Request-Id: f12ab95f-de54-4a6e-59a0-b822df8eb08c::0b4f34cf-e106-429e-a774-c03c396e015e^M
^M
{
  "code": 10001,
  "description": "The service broker returned an invalid response for the request to https://hsdp-rds.app.pivotal.blackbaudcloud.com/v2/service_instances/7374f26c-e4a1-4be8-a57f-db4725255e61?accepts_incomplete=true. Status Code: 500 Internal Server Error, Body: \n    <!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n    <html>\n        <head>\n            <title>Error: 500 Internal Server Error</title>\n            <style type=\"text/css\">\n              html {background-color: #eee; font-family: sans;}\n              body {background-color: #fff; border: 1px solid #ddd;\n                    padding: 15px; margin: 15px;}\n              pre {background-color: #eee; border: 1px solid #ddd; padding: 5px;}\n            </style>\n        </head>\n        <body>\n            <h1>Error: 500 Internal Server Error</h1>\n            <p>Sorry, the requested URL <tt>&#039;https://hsdp-rds.app.pivotal.blackbaudcloud.com/v2/service_instances/7374f26c-e4a1-4be8-a57f-db4725255e61?accepts_incomplete=true&#039;</tt>\n               caused an error:</p>\n            <pre>Internal Server Error</pre>\n        </body>\n    </html>\n",
  "error_code": "CF-ServiceBrokerBadResponse",
  "http": {
    "uri": "https://hsdp-rds.app.pivotal.blackbaudcloud.com/v2/service_instances/7374f26c-e4a1-4be8-a57f-db4725255e61?accepts_incomplete=true",
    "method": "PUT",
    "status": 500
  }
}
```

app output:
```
Connected, tailing logs for app hsdp-rds in org bluemoon / space broker as admin...

2015-10-20T16:23:57.30+0000 [RTR/0]      OUT hsdp-rds.app.pivotal.blackbaudcloud.com - [20/10/2015:16:23:57 +0000] "PUT /v2/service_instances/7374f26c-e4a1-4be8-a57f-db4725255e61?accepts_incomplete=true HTTP/1.1" 500 848 "-" "HTTPClient/1.0 (2.5.3.3, ruby 2.1.6 (2015-04-13))" 10.209.76.157:31500 x_forwarded_for:"52.22.117.12" vcap_request_id:19454afc-b361-4e4d-657d-63960c94a913 response_time:0.002562879 app_id:345d5c19-20b7-4013-964c-e34586fc69b4
2015-10-20T16:23:57.31+0000 [App/0]      ERR Traceback (most recent call last):
2015-10-20T16:23:57.31+0000 [App/0]      ERR   File "/app/.heroku/python/lib/python2.7/site-packages/bottle.py", line 862, in _handle
2015-10-20T16:23:57.31+0000 [App/0]      ERR     return route.call(**args)
2015-10-20T16:23:57.31+0000 [App/0]      ERR   File "/app/.heroku/python/lib/python2.7/site-packages/bottle.py", line 1732, in wrapper
2015-10-20T16:23:57.31+0000 [App/0]      ERR     rv = callback(*a, **ka)
2015-10-20T16:23:57.31+0000 [App/0]      ERR   File "rds_sb_app.py", line 46, in wrapper
2015-10-20T16:23:57.31+0000 [App/0]      ERR     return callback(*args, **kwargs)
2015-10-20T16:23:57.31+0000 [App/0]      ERR   File "/app/.heroku/python/lib/python2.7/site-packages/bottle.py", line 2680, in wrapper
2015-10-20T16:23:57.31+0000 [App/0]      ERR     return func(*a, **ka)
2015-10-20T16:23:57.31+0000 [App/0]      ERR   File "rds_sb_app.py", line 102, in provision
2015-10-20T16:23:57.31+0000 [App/0]      ERR     if plan['id'] == data['plan_id']:
2015-10-20T16:23:57.31+0000 [App/0]      ERR KeyError: 'id'
2015-10-20T16:23:57.31+0000 [App/0]      ERR [pid: 38|app: 0|req: 64/92] 10.209.78.15 () {54 vars in 1124 bytes} [Tue Oct 20 16:23:57 2015] PUT /v2/service_instances/7374f26c-e4a1-4be8-a57f-db4725255e61?accepts_incomplete=true => generated 848 bytes in 1 msecs (HTTP/1.1 500) 2 headers in 99 bytes (1 switches on core 1)
2015-10-20T16:23:59.59+0000 [RTR/0]      OUT hsdp-rds.app.pivotal.blackbaudcloud.com - [20/10/2015:16:23:59 +0000] "DELETE /v2/service_instances/7374f26c-e4a1-4be8-a57f-db4725255e61?plan_id=25079dac-ca5a-4ab3-af1c-24047ba8500d&service_id=e4de7294-4e76-44ac-bf40-08b01282f478 HTTP/1.1" 500 921 "-" "HTTPClient/1.0 (2.5.3.3, ruby 2.1.6 (2015-04-13))" 10.209.76.157:31565 x_forwarded_for:"52.22.117.12" vcap_request_id:28c0e1a6-9581-400e-61bc-f585d547e86e response_time:0.062632286 app_id:345d5c19-20b7-4013-964c-e34586fc69b4
2015-10-20T16:23:59.59+0000 [App/0]      ERR Traceback (most recent call last):
2015-10-20T16:23:59.59+0000 [App/0]      ERR   File "/app/.heroku/python/lib/python2.7/site-packages/bottle.py", line 862, in _handle
2015-10-20T16:23:59.59+0000 [App/0]      ERR     return route.call(**args)
2015-10-20T16:23:59.59+0000 [App/0]      ERR   File "/app/.heroku/python/lib/python2.7/site-packages/bottle.py", line 1732, in wrapper
2015-10-20T16:23:59.59+0000 [App/0]      ERR     rv = callback(*a, **ka)
2015-10-20T16:23:59.59+0000 [App/0]      ERR   File "rds_sb_app.py", line 46, in wrapper
2015-10-20T16:23:59.59+0000 [App/0]      ERR     return callback(*args, **kwargs)
2015-10-20T16:23:59.59+0000 [App/0]      ERR   File "/app/.heroku/python/lib/python2.7/site-packages/bottle.py", line 2680, in wrapper
2015-10-20T16:23:59.59+0000 [App/0]      ERR     return func(*a, **ka)
2015-10-20T16:23:59.59+0000 [App/0]      ERR   File "rds_sb_app.py", line 186, in deprovision
2015-10-20T16:23:59.59+0000 [App/0]      ERR     record = table.get_item(Key={'instance_id': instance_id})['Item']
2015-10-20T16:23:59.59+0000 [App/0]      ERR KeyError: 'Item'
2015-10-20T16:23:59.59+0000 [App/0]      ERR [pid: 37|app: 0|req: 7/93] 10.209.78.15 () {50 vars in 1212 bytes} [Tue Oct 20 16:23:59 2015] DELETE /v2/service_instances/7374f26c-e4a1-4be8-a57f-db4725255e61?plan_id=25079dac-ca5a-4ab3-af1c-24047ba8500d&service_id=e4de7294-4e76-44ac-bf40-08b01282f478 => generated 921 bytes in 61 msecs (HTTP/1.1 500) 2 headers in 99 bytes (1 switches on core 0)
```

**Comment by epoelke** [comment]
@Blackbaud-JasonCrowe I literally just fixed that this morning.  The fix is in the v2 release.  Go ahead and push that version and let me know.  I will leave the issue open until you confirm.

**Comment by Blackbaud-JasonCrowe** [comment]
your fix seem to only deal with part of the issue

**Comment by epoelke** [comment]
@Blackbaud-JasonCrowe do you mean the second error on the missing Key `Item`?

**Comment by Blackbaud-JasonCrowe** [comment]
yes, seems like it deals with a different part of the process (although i don't understand it all yet)

**Comment by Blackbaud-JasonCrowe** [comment]
looks like the fix worked -- i was able to update the app via cf push and then create another service within the same space using a different name.

**Comment by epoelke** [comment]
The second error is due to the the fact that the call to the service broker from the cloud controller returned non-20x http response.  When that happens the cloud controller will immediately send a `deprovision` request to the service broker to clean up any failed provisioning.  Since the instance id never makes it to dynamodb you get the second exception as well.  Its not fatal and really the result of the provision failing at the stage is failed at.  But yeah you are right in that it could be handled better.  I will add a handler for that as well.  The proper way to handle that on the for could foundry is really to return a 410 if the instance id does not exist.

**Comment by Blackbaud-JasonCrowe** [comment]
note: i also confirm creating a service in a different space with the same name works as well. thank you very much! closing issue due to fix from commit 5264bf1d468b4ab0e6b85a2f75c6e031bba771c5
